File: day8_Two_factor/cpp/part1.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <vector>
#include <regex>
#include <chrono>
//Why can't you string switch!! lol

using namespace std;
using namespace std::chrono;

void run(){
    ifstream txtFile("input.txt");
    if(txtFile.is_open()){
        int num = 0;
        int x;
        int y;
        string line;
        string section;
        smatch match;
        vector<vector<int>> grid(6, vector<int>(50, 0));
        regex rectRGX("rect (\\d+)x(\\d+)");
        regex rotateRGX("rotate (column|row) (x|y)=(\\d+) by (\\d+)");
        int tempInt;
        while(txtFile.good()){
                // Line editting
                getline(txtFile, line);
                if (regex_match (line, match, rectRGX)){
                    x = stoi(match.str(1));
                    y = stoi(match.str(2));
                    for (int i = 0; i < x; i++){
                        for (int j = 0; j < y; j++){
                            grid[j][i] = 1;
                        }
                    }
                } else if (regex_match (line, match, rotateRGX)){
                    if (match.str(1).compare("column") == 0){
                        x = stoi(match.str(3));
                        int n = 0;
                        while (n < stoi(match.str(4))){
                            tempInt = grid[5][x];
                            for (int i = 4; i > -1; --i){
                                grid[i+1][x] = grid[i][x];
                            }
                            grid[0][x] = tempInt;
                            n++;
                        }
                    }
                    if (match.str(1).compare("row") == 0){
                        y = stoi(match.str(3));
                        int n = 0;
                        while (n < stoi(match.str(4))){
                            int t_count = 0;
                            tempInt = grid[y][49];
                            for (int i = 48; i > -1; --i){
                                grid[y][i+1] = grid[y][i];
                            }
                            grid[y][0] = tempInt;
                            n++;
                        }
                    }
                }
                section = "";
            }
        txtFile.close();
        for (int i = 0; i < 6; i++){
            for (int j = 0; j < 50; j++){
                if (grid[i][j] == 1){
                    cout << "A";
                    num++;
                } else {
                    cout << " ";
                }
            }
        cout << endl;
        }
    }
}

int main()
{
    high_resolution_clock::time_point t1 = high_resolution_clock::now();
    run();
    high_resolution_clock::time_point t2 = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>( t2 - t1 ).count();

    cout << duration << endl;
    return 0;
}

File: day8_Two_factor/python/part1.py
import re
import os
from time import sleep

def rect(line, grid):
    m = re.search(r"rect (\d+)x(\d+)", line.strip())
    for i in range(int(m.group(2))):
        for j in range(int(m.group(1))):
            grid[i][j] = "#"
    return grid

def rotate(line, grid):
    m = re.match(r"rotate (column|row) (x|y)=(\d+) by (\d+)", line.strip())
    if m.group(2) == "y":
        for _ in range(int(m.group(4))):
            row = int(m.group(3))
            last_char = grid[row][-1]
            for i in range(len(grid[row])-1, 0, -1):
                grid[row][i] = grid[row][i-1]
            grid[row][0] = last_char
    elif m.group(2) == "x":
        for _ in range(int(m.group(4))):
            col = int(m.group(3))
            last_char = grid[-1][col]
            for i in range(len(grid)-1, 0, -1):
                grid[i][col] = grid[i-1][col]
            grid[0][col] = last_char
    else:
        raise(ValueError("Missing x/y"))
    return grid

def display(grid):
    os.system("clear")
    for l in grid:
        print("".join(i for i in l))

def run(filename = "input.txt"):
    grid = [['.' for x in range(50)] for i in range(6)]
    with open(filename) as f:
        for line in f.readlines():
            if "rect" in line:
                grid = rect(line, grid)
            else:
                grid = rotate(line, grid)
            # display(grid)
            # sleep(0.05)
    n = 0
    for l in grid:
        for i in l:
            if i == '#':
                n += 1
    return (n, grid)

if __name__ == "__main__":

    print(run()[0])
